src/measure_theory/measure_space.lean

```diff
@@ -1417,10 +1424,68 @@ lemma finite_at_filter_of_finite (μ : measure α) [finite_measure μ] (f : filt
 lemma measure.finite_at_bot (μ : measure α) : μ.finite_at_filter ⊥ :=
 ⟨∅, mem_bot_sets, by simp only [measure_empty, with_top.zero_lt_top]⟩
 
+/-- A measure `μ` is called σ-finite if there is a countable collection of sets
+  `{ A i | i ∈ ℕ }` such that `μ (A i) < ⊤` and `⋃ i, A i = s`. -/
+class sigma_finite (μ : measure α) : Prop :=
+(exists_finite_spanning_sets :
+  ∃ s : ℕ → set α,
+  (∀ i, is_measurable (s i)) ∧
+  (∀ i, μ (s i) < ⊤) ∧
+  (⋃ i, s i) = univ)
+
+lemma exists_finite_spanning_sets (μ : measure α) [sigma_finite μ] :
+  ∃ s : ℕ → set α,
+  (∀ i, is_measurable (s i)) ∧
+  (∀ i, μ (s i) < ⊤) ∧
+  (⋃ i, s i) = univ :=
+sigma_finite.exists_finite_spanning_sets
+
+/-- A noncomputable way to get a monotone collection of sets that span `univ` and have finite
+  measure using `classical.some`. This definition satisfies monotonicity in addition to all other
+  properties in `sigma_finite`. -/
+def spanning_sets (μ : measure α) [sigma_finite μ] (i : ℕ) : set α :=
+accumulate (classical.some (exists_finite_spanning_sets μ)) i
+
+lemma monotone_spanning_sets (μ : measure α) [sigma_finite μ] :
+  monotone (spanning_sets μ) :=
+monotone_accumulate
+
+lemma is_measurable_spanning_sets (μ : measure α) [sigma_finite μ] (i : ℕ) :
+  is_measurable (spanning_sets μ i) :=
+is_measurable.Union $ λ j, is_measurable.Union_Prop $
+  λ hij, (classical.some_spec (exists_finite_spanning_sets μ)).1 j
+
+lemma measure_spanning_sets_lt_top (μ : measure α) [sigma_finite μ] (i : ℕ) :
+  μ (spanning_sets μ i) < ⊤ :=
+begin
+  rw [spanning_sets, accumulate_nat],
+  apply (measure_bUnion_le _ _).trans_lt,
+  refine (finset.tsum_subtype _ (λ i, μ _)).le.trans_lt _,
+  rw ennreal.sum_lt_top_iff,
+  exact λ j _, (classical.some_spec (exists_finite_spanning_sets μ)).2.1 j,
+  exact (finset.range (i+1)).countable_to_set
+end
 
-instance restrict.finite_measure (μ : measure α) {s : set α} [hs : fact (μ s < ⊤)] :
-  finite_measure (μ.restrict s) :=
-⟨by simp [hs.elim]⟩
+lemma Union_spanning_sets (μ : measure α) [sigma_finite μ] :
+  (⋃ i : ℕ, spanning_sets μ i) = univ :=
+by simp_rw [spanning_sets, Union_accumulate,
+  (classical.some_spec (exists_finite_spanning_sets μ)).2.2]
+
+namespace measure
+lemma supr_restrict_spanning_sets {μ : measure α} [sigma_finite μ] {s : set α}
+  (hs : is_measurable s) :
+  (⨆ i, μ.restrict (spanning_sets μ i) s) = μ s :=
```

Comment:
I'd add `supr_measure_spanning_sets : (⨆ i, μ (spanning_sets μ i)) = μ s`, then used `restrict_apply_univ` for this lemma.

Reply:
I don't quite understand. Did you mean `(⨆ i, μ (spanning_sets μ i)) = μ univ`? But I don't think that proves `supr_restrict_spanning_sets`: it will have as LHS `(⨆ i, μ.restrict (spanning_sets (μ.restrict _) i) s)` instead of `(⨆ i, μ.restrict (spanning_sets μ i) s)`.

That does remind me, let me add an instance that `restrict` preserves sigma-finiteness.

Reply:
You're right.